yarn-project/noir-compiler/src/noir_artifact.ts

```diff
@@ -82,13 +112,56 @@ export interface NoirDebugMetadata {
 /**
  * The compilation artifacts of a given contract.
  */
-export interface NoirCompilationArtifacts {
+export interface NoirContractCompilationArtifacts {
   /**
    * The compiled contract.
    */
   contract: NoirCompiledContract;
+
   /**
    * The artifact that contains the debug metadata about the contract.
    */
   debug?: NoirDebugMetadata;
 }
+
+/**
+ * The compilation artifacts of a given program.
+ */
+export interface NoirProgramCompilationArtifacts {
+  /**
+   * not part of the compilation output, injected later
+   */
+  name: string;
+  /**
+   * The compiled contract.
+   */
+  program: NoirCompiledCircuit;
+
+  /**
+   * The artifact that contains the debug metadata about the contract.
+   */
+  debug?: NoirDebugMetadata;
+}
+
+/**
+ * output of Noir Wasm compilation, can be for a contract or lib/binary
+ */
+export type NoirCompilationResult = NoirContractCompilationArtifacts | NoirProgramCompilationArtifacts;
+
+/**
+ * Check if it has Contract unique property
+ */
+export function isNoirContractCompilationArtifacts(
+  artifact: NoirCompilationResult,
+): artifact is NoirContractCompilationArtifacts {
+  return (artifact as NoirContractCompilationArtifacts).contract !== undefined;
+}
```

Comment:
first time trying to do type matching in typescript, so this may not be best way

Reply:
The best way is a discriminated union, but if the relationship doesn't have a tag field made for discriminated union this can be the best you can do